src/Utils/SimpleScheduler.cpp

```diff
@@ -0,0 +1,265 @@
+        return *error;  // do nothing
+      }
+      auto actualTrack = std::get<QueuedTrack>(actualTrackRet);
+
+      auto setTrackRet = mMusicBackend->setPlayback(actualTrack);
+      if (setTrackRet.has_value()) {
+        LOG(ERROR) << "SimpleScheduler: "
+                   << setTrackRet.value().getErrorMessage();
+        mSchedulerState = Idle;
+        return setTrackRet.value();  // do nothing
+      }
+
+      mSchedulerState = CheckPlaying;
+
+    } break;
+
+    case CheckPlaying: {
+      VLOG(100) << "SimpleScheduler: CheckPlaying";
+
+      auto isPlayingRet = isTrackPlaying(playbackTrackOpt);
+      if (auto error = std::get_if<Error>(&isPlayingRet)) {
+        mSchedulerState = Idle;
+        return *error;
+      }
+      bool isPlaying = std::get<bool>(isPlayingRet);
+
+      if (isPlaying) {
+        mSchedulerState = Playing;
+      }
+    } break;
+
+    case Playing: {
+      VLOG(100) << "SimpleScheduler: Playing";
+      if (!playbackTrackOpt.has_value()) {
+        mSchedulerState = Idle;
+        return Error(ErrorCode::InvalidValue,
+                     "Someone closed Spotify Player on Playback Error");
+      }
+      auto trackFinishedRet = isTrackFinished(playbackTrackOpt);
+      if (auto error = std::get_if<Error>(&trackFinishedRet)) {
+        mSchedulerState = Idle;
+        return *error;
+      }
+      bool trackFinished = std::get<bool>(trackFinishedRet);
+
+      if (trackFinished) {
+        LOG(INFO) << "SimpleScheduler: Track finished";
+        auto emptyValRet = areQueuesEmpty();
+        if (auto error = std::get_if<Error>(&emptyValRet)) {
+          return *error;
+        }
+        bool emptyVal = std::get<bool>(emptyValRet);
+        if (!emptyVal) {
+          mSchedulerState = PlayNextSong;
+        } else {
+          mSchedulerState = Idle;
+        }
+      }
+    } break;
+  }
+  mLastPlaybackTrack = playbackTrackRet;
+
+  return nullopt;
+}
+
+TResult<bool> SimpleScheduler::areQueuesEmpty() {
+  auto adminQueRet = mDataStore->getQueue(QueueType::Admin);
+  if (auto error = std::get_if<Error>(&adminQueRet)) {
+    return *error;
+  }
+  auto adminQueue = std::get<Queue>(adminQueRet);
+  if (!adminQueue.tracks.empty()) {
+    return false;
+  }
+
+  auto normalQueueRet = mDataStore->getQueue(QueueType::Normal);
+  if (auto error = std::get_if<Error>(&normalQueueRet)) {
+    return *error;
+  }
+  auto normalQueue = std::get<Queue>(normalQueueRet);
+  if (!normalQueue.tracks.empty()) {
+    return false;
+  }
+
+  return true;
+}
+
+TResult<bool> SimpleScheduler::isTrackPlaying(
+    std::optional<PlaybackTrack> const &currentOpt) {
+  auto playingTrackRet = mDataStore->getPlayingTrack();
+  if (auto error = std::get_if<Error>(&playingTrackRet)) {
+    return *error;  // do nothing
+  }
+
+  if (!currentOpt.has_value()) {
+    return false;
+  }
+
+  auto playingTrack = std::get<QueuedTrack>(playingTrackRet);
+
+  if (!currentOpt.value().isPlaying) {
+    mMusicBackend->play();
```

Comment:
why is this done here?
wouldn't this always override a user who just clicked pause?

Reply:
This function gets only called in the CheckPlaying state. Sure this funtion overrides the clicked pause, but this is only a short period when the track starts. With this I can now be sure if a track starts playing, is playing or finished